mozetl/topline/topline_dashboard.py

```diff
@@ -0,0 +1,197 @@
+    """
+
+    # A dictionary mapping the column names to the select sub-expression. For
+    # most fields, the subexpr ends up being the column name.
+    topline_columns = {name: name for name in topline_schema.names}
+
+    # Bucket results by the top k countries, and move the rest into other.
+    topline_columns['geo'] = (
+        F.when(F.col('geo').isin(list(countries)), F.col('geo'))
+         .otherwise('Other')
+         .alias('geo')
+    )
+
+    # Use the historical dataset's date format, drop the old date column
+    topline_columns['date'] = (
+        F.from_unixtime(
+            F.unix_timestamp(F.col('report_start'), 'yyyyMMdd'),
+            'yyyy-MM-dd')
+        .alias('date')
+    )
+    topline_columns.pop('report_start', None)
+
+    # The set of rows in the topline_summary can be categorized into
+    # two general groups, attributes and aggregates. Attributes are
+    # categorical in nature, while aggregates are numerical.
+    attributes = set(['geo', 'channel', 'os', 'date'])
+    aggregates = set(topline_columns.keys()) - attributes
+
+    # Cube the results. It's difficult to concisely express
+    # partitioning over a single attribute, and cube each subset of
+    # attributes. Instead, ignore rows where the date field is empty,
+    # which equates to ignoring aggregates over multiple dates.
+    cubed_df = (
+        df
+        .select(*topline_columns.values())
+        .cube(*attributes)
+        .agg(*[F.sum(F.col(x)).alias(x) for x in aggregates])
+        .where(F.col('date').isNotNull())
+        .na.fill('all')  # the only fills in string fields
+    )
+
+    # Remove rows where the aggregates fields are all zeroes, since
+    # these take up extra space in the csv file.
+    filtered_df = (
+        cubed_df
+        .withColumn('total', sum([F.col(x) for x in aggregates]))
+        .where(F.col('total') > 0.0)
+    )
+
+    # Generate select subexpressions based on the historical
+    # schema. Columns missing in the topline_summary from the
+    # historical dataset are aggregates, so they are are filled with
+    # zeroes.
+    def default_column(name):
+        return F.lit(0).alias(name)
+
+    column_expr = [name if name in topline_columns else default_column(name)
+                   for name in historical_schema.names]
+    formatted_df = filtered_df.select(*column_expr)
+
+    return formatted_df
+
+
+def write_dashboard_data(df, bucket, prefix, mode):
+    """ Write the dashboard data to a s3 location. """
+
+    # create a temporary directory to dump files into
+    path = tempfile.mkdtemp()
+
+    # write dataframe to working directory
+    df.repartition(1).write.csv(path, header=True, mode="overwrite")
+
+    # find the file that was created
+    filepath = None
+    for name in os.listdir(path):
+        if name.endswith(".csv"):
+            filepath = os.path.join(path, name)
+            break
+
+    # name of the output key
+    key = "{}/v4-{}.csv".format(prefix, mode)
```

Comment:
Let's rename this to "topline-{weekly|monthly}.csv" to avoid conflicts with the old version of the output.